src/crimson/common/config_proxy.h

```diff
@@ -34,25 +35,33 @@ class ConfigProxy : public seastar::peering_sharded_service<ConfigProxy>
   }
 
   // apply changes to all shards
-  // @param func a functor which accepts @c "ConfigValues&", and returns
-  //             a boolean indicating if the values is changed or not
+  // @param func a functor which accepts @c "ConfigValues&"
   template<typename Func>
   seastar::future<> do_change(Func&& func) {
-    auto new_values = seastar::make_lw_shared(*values);
-    func(*new_values);
-    return seastar::do_with(seastar::make_foreign(new_values),
-			    [this](auto& foreign_values) {
-      return container().invoke_on_all([&foreign_values](ConfigProxy& proxy) {
-        return foreign_values.copy().then([&proxy](auto foreign_values) {
-	  proxy.values.reset();
-	  proxy.values = std::move(foreign_values);
-	  proxy.obs_mgr.apply_changes(proxy.values->changed,
-				      proxy, nullptr);
+    return container().invoke_on(values.get_owner_shard(),
+                                 [func = std::move(func)](ConfigProxy& owner) {
+      // always apply the new settings synchronously on the owner shard, to
+      // avoid racings with other do_change() calls in parallel.
+      owner.values.reset(seastar::make_lw_shared(*owner.values));
+      func(*owner.values);
+      // instead of passing `values` by reference to we need to pass a copy of
+      // it, otherwise the `values` instance could be overwritten by succeeding
+      // calls of do_change(), and other shards will be copying a non-existing
+      // `values`.
+      return owner.values.copy().then([&owner](auto copy_values) {
+        return owner.container().invoke_on_others(
+          [&values = copy_values](ConfigProxy& proxy) {
+            return values.copy().then([&proxy](auto foreign_values) {
+              proxy.values.reset();
+              proxy.values = std::move(foreign_values);
+              proxy.obs_mgr.apply_changes(proxy.values->changed,
+                                          proxy, nullptr);
+            });
+          });
+        }).then([&owner] {
```

Comment:
maybe use `.finally()` here? if we're calling config observers, it's possible that they throw